crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -441,15 +717,58 @@ where
         height: Option<Description::Height>,
         changes: Changes,
     ) -> StorageResult<()> {
-        let mut storage_transaction =
-            StorageTransaction::transaction(&self.db, ConflictPolicy::Overwrite, changes);
+        // cumulative_changes_lock_guard is defined to be Some only when the migration is in progress.
```

Comment:
For clarity: do you mean the function `take_migration_changes` should return only `Option<Changes>`?
If we do not care about keeping the migration changes I can remove keeping the lock guard here. 
I also don't need to readd the migration changes later on.

Note that this works under the assumption that `commit_changes` always commit changes at a `height` that is not present already in the underlying db, so I want to validate this first. If this is not the case, then there is the risk of having a stale version from a migration  overwriting the newer version that has been written into the DB by `commit_changes`.

Reply:
> do you mean the function take_migration_changes should return only Option<Changes>?

Yes, it seems should be enough.

> Note that this works under the assumption that commit_changes always commit changes at a height that is not present already in the underlying db, so I want to validate this first. If this is not the case, then there is the risk of having a stale version from a migration overwriting the newer version that has been written into the DB by commit_changes.

Why we can't apply changes from the `store_modifications_history` on top of the changes from migration? In this case  we will not lose changes from the `store_modifications_history`.

If you worry that later migration will override modification history, then it shouldn't happen because you migrate only v1 entries, while `store_modifications_history` uses v2. Only in the case of rollback can you face the situation where migration with old modifications can override new ones. But as I mentioned in another comment, we can handle this situation just during the migration process by tracking what is the maximum height of the migration(and decreasing this height during the rollback event)

Reply:
> Why we can't apply changes from the store_modifications_history on top of the changes from migration? In this case we will not lose changes from the store_modifications_history. 

The scenario I was concerned about goes as follows (As I mention this can happen only if we commit twice at the same height, which does not seem something that can happen without rollbacking first)

- You start the function to commit at a height for which there is a V1 key present already, with a different set of changes. This function takes the migration changes to commit, and find that they are empty
- You migrate the same key from V1 to V2. The changes (which are different from the ones from V1) are persisted in memory
- The commit transaction commits into rocksDB. Later on, a second commit transaction at a different height gets the stale migration changes that have been persisted in memory. 

So if the following above cannot happen then I am convinced that it should be fine to not keep the lock for long